backends/candle/src/models/nomic.rs

```diff
@@ -92,48 +95,336 @@ impl NomicBertEmbeddings {
 }
 
 pub struct NomicBertGatedMLP {
-    gate_up_proj: Linear,
-    down_proj: Linear,
+    fc11: Linear,
+    fc12: Linear,
+    fc2: Linear,
+    activation: HiddenAct,
 
     span: tracing::Span,
 }
 
 impl NomicBertGatedMLP {
     pub fn load(vb: VarBuilder, config: &NomicConfig) -> Result<Self> {
         let intermediate_size = config.n_inner;
+        let activation = config.activation_function.clone();
 
-        let gate_proj_weight = vb
-            .pp("fc12")
+        let fc11_weight = vb
+            .pp("fc11")
             .get((intermediate_size, config.n_embd), "weight")?;
+        let fc11_bias = if config.mlp_fc1_bias {
+            Some(vb.pp("fc11").get((intermediate_size,), "bias")?)
+        } else {
+            None
+        };
+        let fc11 = Linear::new(fc11_weight, fc11_bias, None);
 
-        let up_proj_weight = vb
-            .pp("fc11")
+        let fc12_weight = vb
+            .pp("fc12")
             .get((intermediate_size, config.n_embd), "weight")?;
+        let fc12_bias = if config.mlp_fc1_bias {
+            Some(vb.pp("fc12").get((intermediate_size,), "bias")?)
+        } else {
+            None
+        };
+        let fc12 = Linear::new(fc12_weight, fc12_bias, None);
+
+        let fc2_weight = vb
+            .pp("fc2")
+            .get((config.n_embd, intermediate_size), "weight")?;
+        let fc2_bias = if config.mlp_fc2_bias {
+            Some(vb.pp("fc2").get((config.n_embd,), "bias")?)
+        } else {
+            None
+        };
+        let fc2 = Linear::new(fc2_weight, fc2_bias, None);
+
+        Ok(Self {
+            fc11,
+            fc12,
+            fc2,
+            activation,
+            span: tracing::span!(tracing::Level::TRACE, "gated_mlp"),
+        })
+    }
+
+    pub fn forward(&self, hidden_states: &Tensor) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let y = self.fc11.forward(hidden_states)?;
+        let gate = self.fc12.forward(hidden_states)?;
+
+        let activated_gate = match self.activation {
+            HiddenAct::Gelu => gate.gelu()?,
+            HiddenAct::Swiglu => gate.silu()?,
+            _ => candle_nn::ops::sigmoid(&gate)?,
```

Comment:
This seems wrong, HiddenAct::Sigmoid should be the catch, we should handle, or panic for other activations.

Given the simplicity of activations, we could definitely abstract it away into a layer that is common to all models.

Reply:
I've just implemented a `forward()` method for the `HiddenAct` enum to forward the activation with the given input. https://github.com/huggingface/text-embeddings-inference/pull/596/commits/188d0983d2cb9fca62850d8a10b2e21a0ef5e8da

Would this kind of approach be good?